man/putdata.3
.\" putdata.3.  The putdata man page.
.\"
.\" (C) 2008, 2009 D. V. Wiebe
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\"
.\" This file is part of the GetData project.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.2 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, with no Front-Cover Texts, and with no Back-Cover
.\" Texts.  A copy of the license is included in the `COPYING.DOC' file
.\" as part of this distribution.
.\"
.TH putdata 3 "19 October 2009" "Version 0.6.0" "GETDATA"
.SH NAME
putdata \(em write data to a dirfile database
.SH SYNOPSIS
.B #include <getdata.h>
.HP
.nh
.ad l
.BI "size_t putdata(DIRFILE *" dirfile ", const char *" field_code ", off_t"
.IB first_frame ", off_t " first_sample ", size_t " num_frames ", size_t"
.IB num_samples ", gd_type_t " data_type ", const void *" data_in );
.hy
.ad n
.SH DESCRIPTION
The
.BR putdata ()
function writes data to a dirfile(5) database specified by
.I dirfile
for the field
.IR field_code ,
which may not contain a representation suffix.  It writes
.I num_frames
frames plus
.I num_samples
samples to this field, starting 
.I first_sample
samples past frame
.IR first_frame . 
The data is read from the user-supplied buffer
.IR data_in,
which is has a data type specified by
.IR data_type .
This interface cannot write to field representations.

The 
.I dirfile
argument must point to a valid DIRFILE object previously created by a call to
.BR dirfile_open (3).

The first sample written will be
.RS
.IR first_frame " * " samples_per_frame " + " first_sample
.RE
as measured from the start of the dirfile, where
.I samples_per_frame
is the number of samples per frame as returned by
.BR get_spf (3).
The number of samples which
.BR putdata ()
attempts to write is, similarly,
.RS
.IR num_frames " * " samples_per_frame " + " num_samples .
.RE
Although calling
.BR putdata ()
using both samples and frames is possible, the function is typically called
with either
.IR num_samples " and " first_sample ,
or
.IR num_frames " and " first_frames ,
equal to zero.

The 
.I data_type
argument should be one of the following symbols, which indicates the type of
the input data:
.RS
.TP 11
.I GD_UINT8
unsigned 8-bit integer
.TP
.I GD_INT8
signed (two's complement) 8-bit integer
.TP
.I GD_UINT16
unsigned 16-bit integer
.TP
.I GD_INT16
signed (two's complement) 16-bit integer
.TP
.I GD_UINT32
unsigned 32-bit integer
.TP
.I GD_INT32
signed (two's complement) 32-bit integer
.TP
.I GD_UINT64
unsigned 64-bit integer
.TP
.I GD_INT64
signed (two's complement) 64-bit integer
.TP
.IR GD_FLOAT32 \~or\~ GD_FLOAT
IEEE-754 standard 32-bit single precision floating point number
.TP
.IR GD_FLOAT64 \~or\~ GD_DOUBLE
IEEE-754 standard 64-bit double precision floating point number
.RE

The type of the input data need not be the same as the type of the data stored
in the database.  Type conversion will be performed as necessary to write the
appropriate type.  The argument
.I data_in
must point to a valid memory location of containing all the data to be written.
.SH RETURN VALUE
In all cases,
.BR putdata ()
returns the number of samples (not bytes) successfully written to the database,
which may be zero if an error has occurred.

If an error has occurred, the dirfile error
will be set to a non-zero value.  Possible error values are:
.TP 8
.B GD_E_ACCMODE
The specified
.I dirfile
was opened read-only.
.TP
.B GD_E_ALLOC
The library was unable to allocate memory.
.TP
.B GD_E_BAD_CODE
The field specified by
.IR field_code ,
or one of the fields it uses for input, was not found in the database.
.TP
.B GD_E_BAD_DIRFILE
An invalid
.I dirfile
was supplied.
.TP
.B GD_E_BAD_FIELD_TYPE
Either the field specified by
.IR field_code ,
or one of the fields it uses for input, was of
.B MULTIPLY
type, or
.B LINCOM
type with more than one input fields.  In this case,
.BR putdata ()
has no knowledge on how to partition the input data.  Alternately, the caller
may have attempted to write to the implicit INDEX field, which is not possible.
.TP
.B GD_E_BAD_REPR
The representation suffix specified in
.I field_code
was not recognised, or an attempt was made to write to a field representation,
instead of the underlying field.
.TP
.B GD_E_BAD_TYPE
An invalid
.I data_type
was specified.
.TP
.B GD_E_INTERNAL_ERROR
An internal error occurred in the library while trying to perform the task.
This indicates a bug in the library.  Please report the incident to the
maintainer.
.TP
.B GD_E_OPEN_LINFILE
An error occurred while trying to read a LINTERP table from disk.
.TP
.B GD_E_PROTECTED
The data of the
.B RAW
field backing
.I field_code
was protected from change by a PROTECT directive.
.TP
.B GD_E_RANGE
An attempt was made to write data before the beginning-of-frame marker for
.IR field_code ,
or the raw field it depends on.
.TP
.B GD_E_RAW_IO
An error occurred while trying to open, read from, or write to a file on disk
containing a raw field.
.TP
.B GD_E_RECURSE_LEVEL
Too many levels of recursion were encountered while trying to resolve
.IR field_code .
This usually indicates a circular dependency in field specification in the
dirfile.
.TP
.B GD_E_UNSUPPORTED
Reading from dirfiles with the encoding scheme of the specified dirfile is not
supported by the library.  See
.BR dirfile-encoding (5)
for details on dirfile encoding schemes.
.RE
.P
The dirfile error may be retrieved by calling
.BR get_error (3).
A descriptive error string for the last error encountered can be obtained from
a call to
.BR get_error_string (3).
.SH SEE ALSO
.BR dirfile (5),
.BR dirfile-encoding (5),
.BR dirfile_open (3),
.BR get_error (3),
.BR get_error_string (3),
.BR get_spf (3)


NEWS
New in version 0.6.0rc3:

  License

  * The GetData library and its bindings are now distributed under the GNU
    Lesser Public License, version 2.1 or later.

  Dirfile Changes

  * GetData now support Dirfile Standards Version 7.  Standards Version 7 adds
    complex data types, a signed bitfield (SBIT), a polynomial derived field
    (POLYNOM), and other features detailed below.
  * Field codes can now contain a "representation suffix", which allows
    specifying a complex norm used to convert complex valued data into purely
    real data.
  * The n_fields parameter in a LINCOM specification is now optional.
  * A new encoding scheme which supports LZMA compressed data using the .xz
    container format has been added.
  * Look-up tables may contain a complex y-value.  This prohibits writing to
    the associated LINTERP field.

  Library Changes

  * The field parameters spf (RAW), bitnum and numbits (BIT and SBIT) are now
    internally restricted to 16-bit signed numbers, which should be more than
    enough.  The public API still calls these 'int', but callers should never
    assume this means more than 16-bits.
  * The shift field parameter is now a long int, which ensures at least 32-bits.
  * The dirfile flags, including encoding and endianness symbols are now
    explicitly long ints.  Previously these quantities mixed long int with int
    arbitrarily.  This affects the public API.
  * The parser callback has received a new, API breaking signature, which should
    hopefully reduce the amount of API breakage in the future.  It is also now
    possible for the caller to pass data through GetData to the callback, if
    desired.
  * Computation of derived fields can now handle complex valued inputs, and
    computation will occur in the complex plane when required.
  * The computation of LINCOM fields has been re-written and is now ~20% faster
    for LINCOMs with 2 or 3 input fields.  (No change for LINCOMs of 1 field.)
  * A new open flag, GD_PRETTY_PRINT, tells the library to attempt to write a
    nicer looking format file.  What this specifically means, isn't part of the
    public API and shouldn't be relied on in portable code.
  * BUG FIX: dirfile_madd_spec() and dirfile_add_spec() didn't properly flag
    the fragments they modified as dirty, leading to possible loss of metadata.
  * BUG FIX: calling get_nframes on a bzip2 encoded dirfile would report the
    wrong dirfile length.
  * BUG FIX: LINTERP tables are now closed after loading, resolving a file
    descriptor leak of great age (ie. pre-0.3.0).
  * BUG FIX: When re-writing a format file fragment containing fields with
    CONST scalar field parameters, GetData no longer replaces the CONST field
    codes in the field specification lines with the value of the CONST field.
  * BUG FIX: when moving a field with dirfile_move(), the fragment index in
    the entry objects of metafields defined for the field were not updated.
    (The metafields were placed in the correct fragment when the metadata was
    flushed to disk, despite this.)
  * BUG FIX: Calling getdata() or putdata() on 64-bit architectures in certain
    situations involving a non-zero frameoffset no longer results in bizarrity.

  API Changes

  * This release breaks ABI compatibility and, to a lesser extent, API
    compatibility.  The SOVERSION of the library has been incremented
    accordingly.
  * An array containing the CONST scalar field codes used when non-literal
    parameters are used in the specification of various different fields has
    been added to the gd_entry_t object.  As a side-effect of this,
    dirfile_add(), dirfile_alter_entry(), and dirfile_madd() can be used to
    manipulate non-literal field parameters.
  * A function, get_native_type(), now exists which returns the underlying data
    type of a specified field code.
  * A function, get_framenum(), is now available to perform a "reverse look-up"
    on a field: given a data value, it reports the frame number where the field
    has that value.  The field must be monotonic.
  * A function, dirfile_validate(), now exists which can be used to check the
    validity of a field code before use.
  * Two new error codes have been added: GD_E_DOMAIN, potentially returned by
    get_framenum(), and GD_E_BAD_REPR, indicating an unrecognised representation
    suffix, or an attempt to write to a representation.
  * A new syntax error suberror code has been added: GD_E_FORMAT_LITERAL, which
    indicates a syntax error in a complex literal number.
  * BUG FIX: On success, dirfilename(), dirfile_reference(), and get_reference()
    didn't change the error code to GD_E_OK, resulting in spurious error codes
    from these functions.
  * BUG FIX: dirfile_madd_string() and dirfile_madd_const() now don't fail
    erroneously by reporting GD_E_BAD_CODE.
  * BUG FIX: error strings relating to GD_E_DELETE error codes now report the
    proper error information.
  * BUG FIX: the list returned by get_mstring() is now properly NULL terminated.
  * BUG FIX: the entry->table member now reports the proper path to the LUT.
    (That is, the path stored in the format file.)  Previously it incorrectly
    reported a path relative to the base dirfile directory.

  Legacy API Changes

  * Because they would otherwise be unreported, in the structure returned by
    GetFormat(), POLYNOM entries are reported as LINCOM entries, by discarding
    higher order terms.  Similarly, SBIT entries are reported as BIT entries.

  Bindings

  * Bindings for Python and the Interactive Data Language (IDL) have been
    added.
  * Bindings for the functionality introduced into 0.6.0 have been added.
  * C++: The missing bindings for dirfilename(), get_nvectors(), and
    get_fragment_index() have been added.
  * C++: Empty entry class constructors have been added which allow delayed
    specification of field metadata.
  * C++: The dirfile.h header now includes all other C++ headers, and is now

src/putdata.c
/* (C) 2003-2005 C. Barth Netterfield
 * (C) 2003-2005 Theodore Kisner
 * (C) 2005-2009 D. V. Wiebe
 *
 ***************************************************************************
 *
 * This file is part of the GetData project.
 *
 * GetData is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * GetData is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with GetData; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "internal.h"

#ifdef STDC_HEADERS
#include <inttypes.h>
#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <stdlib.h>
#endif

static size_t _GD_DoRawOut(DIRFILE *D, gd_entry_t *E, off64_t s0,
    size_t ns, gd_type_t data_type, const void *data_in)
{
  size_t n_wrote;
  void *databuffer;

  /* check protection */
  if (D->fragment[E->fragment_index].protection & GD_PROTECT_DATA) {
    _GD_SetError(D, GD_E_PROTECTED, GD_E_PROTECTED_DATA, NULL, 0,
        D->fragment[E->fragment_index].cname);
    dreturn("%i", 0);
    return 0;
  }

  dtrace("%p, %p, %lli, %zi, 0x%x, %p", D, E, s0, ns, data_type, data_in);

  if (s0 < D->fragment[E->fragment_index].frame_offset * E->spf) {
    _GD_SetError(D, GD_E_RANGE, GD_E_OUT_OF_RANGE, NULL, 0, NULL);
    dreturn("%zi", 0);
    return 0;
  }

  s0 -= D->fragment[E->fragment_index].frame_offset * E->spf;

  databuffer = _GD_Alloc(D, E->data_type, ns);

  if (databuffer == NULL) {
    dreturn("%zi", 0);
    return 0;
  }

  _GD_ConvertType(D, data_in, data_type, databuffer, E->data_type, ns);

  if (D->error) { /* bad input type */
    free(databuffer);
    dreturn("%zi", 0);
    return 0;
  }

  if (!_GD_Supports(D, E, GD_EF_OPEN | GD_EF_SEEK | GD_EF_WRITE)) {
    dreturn("%i", 0);
    return 0;
  }

  if (_gd_ef[E->e->file[0].encoding].ecor &&
      (D->fragment[E->fragment_index].byte_sex ==
#ifdef WORDS_BIGENDIAN
       GD_LITTLE_ENDIAN
#else
       GD_BIG_ENDIAN
#endif
      ))
  {
    if (E->data_type & GD_COMPLEX)
      _GD_FixEndianness(databuffer, E->e->size / 2, ns * 2);
    else
      _GD_FixEndianness(databuffer, E->e->size, ns);
  }

  /* write data to file. */

  if (E->e->file[0].fp < 0) {
    /* open file for reading / writing if not already opened */

    if (_GD_SetEncodedName(D, E->e->file, E->e->filebase, 0)) {
      dreturn("%i", 0);
      return 0;
    } else if ((*_gd_ef[E->e->file[0].encoding].open)(E->e->file,
          D->flags & GD_ACCMODE, 1))
    {
      _GD_SetError(D, GD_E_RAW_IO, 0, E->e->file[0].name, errno, NULL);
      dreturn("%zi", 0);
      return 0;
    }
  }

  if ((*_gd_ef[E->e->file[0].encoding].seek)(E->e->file, s0, E->data_type, 1)
      == -1)
  {
      _GD_SetError(D, GD_E_RAW_IO, 0, E->e->file[0].name, errno, NULL);
      dreturn("%zi", 0);
      return 0;
  }

  n_wrote = (*_gd_ef[E->e->file[0].encoding].write)(E->e->file, databuffer,
      E->data_type, ns);

  free(databuffer);

  dreturn("%zi", n_wrote);
  return n_wrote;
}

static size_t _GD_DoLinterpOut(DIRFILE* D, gd_entry_t *E, off64_t first_samp,
    size_t num_samp, gd_type_t data_type, const void *data_in)
{
  size_t n_wrote;

  dtrace("%p, %p, %lli, %zi, 0x%x, %p", D, E, first_samp, num_samp, data_type,
      data_in);

  /* if the table is complex valued, we can't invert it */
  if (E->e->complex_table) {
    _GD_SetError(D, GD_E_DOMAIN, GD_E_DOMAIN_COMPLEX, NULL, 0, NULL);
    dreturn("%zi", 0);
    return 0;
  }

  if (E->e->table_len < 0) {
    _GD_ReadLinterpFile(D, E);
    if (D->error != GD_E_OK) {
      dreturn("%zi", 0);
      return 0;
    }
  }

  /* Interpolate X(y) instead of Y(x) */
  if (_GD_BadInput(D, E, 0)) {
    dreturn("%zi", 0);
    return 0;
  }

  double *tmpbuf = _GD_Alloc(D, GD_FLOAT64, num_samp);
  if (tmpbuf == NULL) {
    free(tmpbuf);
    dreturn("%zi", 0);
    return 0;
  }

  _GD_ConvertType(D, data_in, data_type, tmpbuf, GD_FLOAT64, num_samp);

  if (D->error) {
    free(tmpbuf);
    dreturn("%zi", 0);
    return 0;
  }

  _GD_LinterpData(D, tmpbuf, GD_FLOAT64, tmpbuf, num_samp, E->e->y, E->e->x,
      E->e->table_len);

  if (D->error != GD_E_OK) {
    free(tmpbuf);
    dreturn("%zi", 0);
    return 0;
  }

  n_wrote = _GD_DoFieldOut(D, E->e->entry[0], E->e->repr[0], first_samp,
      num_samp, GD_FLOAT64, tmpbuf);

  free(tmpbuf);

  dreturn("%zi", n_wrote);
  return n_wrote;
}

static size_t _GD_DoLincomOut(DIRFILE* D, gd_entry_t *E, off64_t first_samp,
    size_t num_samp, gd_type_t data_type, const void *data_in)
{
  size_t n_wrote;
  void* tmpbuf;

  dtrace("%p, %p, %lli, %zi, 0x%x, %p", D, E, first_samp, num_samp, data_type,
      data_in);

  /* we cannot write to LINCOM fields that are a linear combination */
  /* of more than one raw field (no way to know how to split data). */

  if (E->n_fields > 1) {
    _GD_SetError(D, GD_E_BAD_FIELD_TYPE, GD_E_FIELD_PUT, NULL, 0, E->field);
    dreturn("%zi", 0);
    return 0;
  }

  if (_GD_BadInput(D, E, 0)) {
    dreturn("%zi", 0);
    return 0;
  }

  /* do the inverse scaling */
  if (D->error != GD_E_OK) {
    dreturn("%zi", 0);
    return 0;
  }

  /* writeable copy */
  tmpbuf = _GD_Alloc(D, data_type, num_samp);

  if (tmpbuf == NULL) {
    dreturn("%zi", 0);
    return 0;
  }

  memcpy(tmpbuf, data_in, num_samp * GD_SIZE(data_type));

  if (E->comp_scal) {
    double complex cm = 1 / E->cm[0];
    double complex cb = -E->cb[0] / E->cm[0];
    _GD_CLincomData(D, 1, tmpbuf, data_type, NULL, NULL, &cm, &cb, NULL,
        num_samp);
  } else {
    double m = 1 / E->m[0];
    double b = -E->b[0] / E->m[0];
    _GD_LincomData(D, 1, tmpbuf, data_type, NULL, NULL, &m, &b, NULL, num_samp);
  }

  if (D->error != GD_E_OK) {
    free(tmpbuf);
    dreturn("%zi", 0);
    return 0;
  }

  n_wrote = _GD_DoFieldOut(D, E->e->entry[0], E->e->repr[0], first_samp,
      num_samp, data_type, tmpbuf);
  free(tmpbuf);

  dreturn("%zi", n_wrote);
  return n_wrote;
}

static size_t _GD_DoBitOut(DIRFILE* D, gd_entry_t *E, off64_t first_samp,
    size_t num_samp, gd_type_t data_type, const void *data_in)
{
  uint64_t *tmpbuf;
  uint64_t *readbuf;
  size_t i, n_wrote;

  dtrace("%p, %p, %lli, %zi, 0x%x, %p", D, E, first_samp, num_samp, data_type,
      data_in);

  const uint64_t mask = (E->numbits == 64) ? 0xffffffffffffffffULL :
    ((uint64_t)1 << E->numbits) - 1;

  if (_GD_BadInput(D, E, 0)) {
    dreturn("%zi", 0);
    return 0;
  }

  tmpbuf = _GD_Alloc(D, GD_UINT64, num_samp);
  readbuf = _GD_Alloc(D, GD_UINT64, num_samp);

  if (tmpbuf == NULL || readbuf == NULL) {
    free(tmpbuf);
    free(readbuf);
    dreturn("%zi", 0);
    return 0;
  }

  memset(tmpbuf, 0, sizeof(uint64_t) * num_samp);
  memset(readbuf, 0, sizeof(uint64_t) * num_samp);

  _GD_ConvertType(D, data_in, data_type, (void*)tmpbuf, GD_UINT64, num_samp);

  /* first, READ the field in so that we can change the bits    */
  _GD_DoField(D, E->e->entry[0], E->e->repr[0], first_samp, num_samp, GD_UINT64,
      readbuf);

  /* error encountered, abort */
  if (D->error != GD_E_OK) {
    dreturn("%zi", 0);
    return 0;
  }

  /* now go through and set the correct bits in each field value */
  for (i = 0; i < num_samp; i++)
    readbuf[i] = (readbuf[i] & ~(mask << E->bitnum)) |
      (tmpbuf[i] & mask) << E->bitnum;

  /* write the modified data out */
  n_wrote = _GD_DoFieldOut(D, E->e->entry[0], E->e->repr[0], first_samp,
      num_samp, GD_UINT64, (void*)readbuf);

  free(readbuf);
  free(tmpbuf);

  dreturn("%zi", n_wrote);
  return n_wrote;
}

static size_t _GD_DoPhaseOut(DIRFILE* D, gd_entry_t *E, off64_t first_samp,
    size_t num_samp, gd_type_t data_type, const void *data_in)
{
  size_t n_wrote;

  dtrace("%p, %p, %lli, %zi, 0x%x, %p", D, E, first_samp, num_samp, data_type,
      data_in);

  if (_GD_BadInput(D, E, 0)) {
    dreturn("%i", 0);
    return 0;
  }

  n_wrote = _GD_DoFieldOut(D, E->e->entry[0], E->e->repr[0], first_samp +
      E->shift, num_samp, data_type, data_in);

  dreturn("%zi", n_wrote);

  return n_wrote;
}

static size_t _GD_DoPolynomOut(DIRFILE* D, gd_entry_t *E, off64_t first_samp,
    size_t num_samp, gd_type_t data_type, const void *data_in)
{
  size_t n_wrote;
  void* tmpbuf;

  dtrace("%p, %p, %lli, %zi, 0x%x, %p", D, E, first_samp, num_samp, data_type,
      data_in);

  /* we cannot write to POLYNOM fields that are quadradic or higher order */

  if (E->poly_ord > 1) {
    _GD_SetError(D, GD_E_BAD_FIELD_TYPE, GD_E_FIELD_PUT, NULL, 0, E->field);
    dreturn("%zi", 0);
    return 0;
  }

  if (_GD_BadInput(D, E, 0)) {
    dreturn("%zi", 0);
    return 0;
  }

  /* do the inverse scaling */
  if (D->error != GD_E_OK) {
    dreturn("%zi", 0);
    return 0;
  }

  /* writeable copy */
  tmpbuf = _GD_Alloc(D, data_type, num_samp);

  if (tmpbuf == NULL) {
    dreturn("%zi", 0);
    return 0;
  }

  memcpy(tmpbuf, data_in, num_samp * GD_SIZE(data_type));

  if (E->comp_scal) {
    double complex cm = 1 / E->ca[1];
    double complex cb = -E->ca[0] / E->ca[1];
    _GD_CLincomData(D, 1, tmpbuf, data_type, NULL, NULL, &cm, &cb, NULL,
        num_samp);
  } else {
    double m = 1 / E->a[1];
    double b = -E->a[0] / E->a[1];
    _GD_LincomData(D, 1, tmpbuf, data_type, NULL, NULL, &m, &b, NULL, num_samp);
  }

  if (D->error != GD_E_OK) {
    free(tmpbuf);
    dreturn("%zi", 0);
    return 0;
  }

  n_wrote = _GD_DoFieldOut(D, E->e->entry[0], E->e->repr[0], first_samp,
      num_samp, data_type, tmpbuf);
  free(tmpbuf);

  dreturn("%zi", n_wrote);
  return n_wrote;
}

static size_t _GD_DoConstOut(DIRFILE* D, gd_entry_t *E, gd_type_t data_type,
    const void *data_in)
{
  dtrace("%p, %p, 0x%x, %p", D, E, data_type, data_in);

  /* check protection */
  if (D->fragment[E->fragment_index].protection & GD_PROTECT_FORMAT)
    _GD_SetError(D, GD_E_PROTECTED, GD_E_PROTECTED_FORMAT, NULL, 0,
        D->fragment[E->fragment_index].cname);
  else if (E->const_type & GD_SIGNED)
    _GD_ConvertType(D, data_in, data_type, &E->e->iconst, GD_INT64, 1);
  else if (E->const_type & GD_COMPLEX)
    _GD_ConvertType(D, data_in, data_type, &E->e->cconst, GD_COMPLEX128, 1);
  else if (E->const_type & GD_IEEE754)
    _GD_ConvertType(D, data_in, data_type, &E->e->dconst, GD_FLOAT64, 1);
  else
    _GD_ConvertType(D, data_in, data_type, &E->e->uconst, GD_UINT64, 1);

  if (D->error) { /* bad input type */
    dreturn("%zi", 0);
    return 0;
  }

  D->fragment[E->fragment_index].modified = 1;

  dreturn("%i", 1);
  return 1;
}

static size_t _GD_DoStringOut(DIRFILE* D, gd_entry_t *E, const char *data_in)
{
  dtrace("%p, %p, %p", D, E, data_in);
  char* ptr = E->e->string;

  /* check protection */
  if (D->fragment[E->fragment_index].protection & GD_PROTECT_FORMAT) {
    _GD_SetError(D, GD_E_PROTECTED, GD_E_PROTECTED_FORMAT, NULL, 0,
        D->fragment[E->fragment_index].cname);
    dreturn("%i", 0);
    return 0;
  }

  E->e->string = strdup(data_in);
  if (E->e->string == NULL) {
    E->e->string = ptr;
    _GD_SetError(D, GD_E_ALLOC, 0, NULL, 0, NULL);
    dreturn("%i", 0);
    return 0;
  }
  free(ptr);
  D->fragment[E->fragment_index].modified = 1;

  dreturn("%i", strlen(E->e->string) + 1);
  return strlen(E->e->string) + 1;
}

size_t _GD_DoFieldOut(DIRFILE *D, gd_entry_t* E, int repr, off64_t first_samp,
    size_t num_samp, gd_type_t data_type, const void *data_in)
{
  size_t n_wrote = 0;

  dtrace("%p, %p, %i, %lli, %zi, 0x%x, %p", D, E, repr, first_samp, num_samp,
      data_type, data_in);

  if (++D->recurse_level >= GD_MAX_RECURSE_LEVEL) {
    _GD_SetError(D, GD_E_RECURSE_LEVEL, 0, NULL, 0, E->field);
    D->recurse_level--;
    dreturn("%zi", 0);
    return 0;
  }

  if (!E->e->calculated)
    _GD_CalculateEntry(D, E);

  if (D->error) {
    dreturn("%i", 0);
    return 0;
  }

  /* writing to representations is prohibited */
  if (repr != GD_REPR_NONE) {
    const char r[2] = {repr, 0};
    _GD_SetError(D, GD_E_BAD_REPR, GD_E_REPR_PUT, NULL, 0, r);
    dreturn("%zi", 0);
    return 0;
  }

  switch (E->field_type) {
    case GD_RAW_ENTRY:
      n_wrote = _GD_DoRawOut(D, E, first_samp, num_samp, data_type, data_in);
      break;
    case GD_LINTERP_ENTRY:
      n_wrote = _GD_DoLinterpOut(D, E, first_samp, num_samp, data_type,
          data_in);
      break;
    case GD_LINCOM_ENTRY:
      n_wrote = _GD_DoLincomOut(D, E, first_samp, num_samp, data_type, data_in);
      break;
    case GD_BIT_ENTRY:
    case GD_SBIT_ENTRY:
      n_wrote = _GD_DoBitOut(D, E, first_samp, num_samp, data_type, data_in);
      break;
    case GD_MULTIPLY_ENTRY:
    case GD_INDEX_ENTRY:
      _GD_SetError(D, GD_E_BAD_FIELD_TYPE, GD_E_FIELD_PUT, NULL, 0, E->field);
      break;
    case GD_PHASE_ENTRY:
      n_wrote = _GD_DoPhaseOut(D, E, first_samp, num_samp, data_type, data_in);
      break;
    case GD_POLYNOM_ENTRY:
      n_wrote = _GD_DoPolynomOut(D, E, first_samp, num_samp, data_type,
          data_in);
      break;
    case GD_CONST_ENTRY:
      n_wrote = _GD_DoConstOut(D, E, data_type, data_in);
      break;
    case GD_STRING_ENTRY:
      n_wrote = _GD_DoStringOut(D, E, data_in);
      break;
    case GD_NO_ENTRY:
      _GD_InternalError(D);
      break;
  }

  D->recurse_level--;
  dreturn("%zi", n_wrote);
  return n_wrote;
}

/* this function is little more than a public boilerplate for _GD_DoFieldOut */
size_t putdata64(DIRFILE* D, const char *field_code_in, off64_t first_frame,
    off64_t first_samp, size_t num_frames, size_t num_samp, gd_type_t data_type,
    const void *data_in)
{
  size_t n_wrote = 0;
  gd_entry_t *entry;
  char* field_code;
  int repr;

  dtrace("%p, \"%s\", %lli, %lli, %zi, %zi, 0x%x, %p", D, field_code_in,
      first_frame, first_samp, num_frames, num_samp, data_type, data_in);

  if (D->flags & GD_INVALID) {/* don't crash */
    _GD_SetError(D, GD_E_BAD_DIRFILE, 0, NULL, 0, NULL);
    dreturn("%zi", 0);
    return 0;
  }

  if ((D->flags & GD_ACCMODE) != GD_RDWR) {
    _GD_SetError(D, GD_E_ACCMODE, 0, NULL, 0, NULL);
    dreturn("%zi", 0);
    return 0;
  }

  _GD_ClearError(D);

  repr = _GD_GetRepr(D, field_code_in, &field_code);

  if (D->error) {
    dreturn("%i", 0);
    return 0;
  }
  entry = _GD_FindField(D, field_code, NULL);

  if (entry == NULL)
    _GD_SetError(D, GD_E_BAD_CODE, 0, NULL, 0, field_code);
  else if (entry->field_type & GD_SCALAR_ENTRY)
    _GD_SetError(D, GD_E_BAD_FIELD_TYPE, GD_E_FIELD_BAD, NULL, 0, field_code);

  if (field_code != field_code_in)
    free(field_code);

  if (D->error) {
      dreturn("%i", 0);
      return 0;
  }

  /* get the samples per frame */
  _gd_spf_t spf = _GD_GetSPF(D, entry);

  if (D->error) {
    dreturn("%i", 0);
    return 0;
  }

  first_samp += spf * first_frame;
  num_samp += spf * num_frames;

  n_wrote = _GD_DoFieldOut(D, entry, repr, first_samp, num_samp, data_type,
      data_in);

  dreturn("%zi", n_wrote);
  return n_wrote;
}

/* 32(ish)-bit wrapper for the 64-bit version, when needed */
size_t putdata(DIRFILE* D, const char *field_code, off_t first_frame,
    off_t first_samp, size_t num_frames, size_t num_samp, gd_type_t data_type,
    const void *data_in)
{
  return putdata64(D, field_code, first_frame, first_samp, num_frames, num_samp,
      data_type, data_in);
}
/* vim: ts=2 sw=2 et
*/
